src/roswire/ros2/state.py

```diff
@@ -14,12 +18,55 @@
 class ROS2StateProbe:
     """Provides an interface for obtaining the instantaneous state of a ROS
     system in terms of its publishers, subscribers, and services."""
+    _app_instance: 'AppInstance' = attr.ib()
+
     @classmethod
     def for_app_instance(cls, app_instance: 'AppInstance') -> 'ROS2StateProbe':
-        raise NotImplementedError
+        return ROS2StateProbe(app_instance=app_instance)
 
     def probe(self) -> SystemState:
         """Obtains the instantaneous state of the associated ROS system."""
-        raise NotImplementedError
+        node_to_state: Dict[Optional[str], Dict[str, List[str]]] = {'pub': {},
+                                                                    'sub': {},
+                                                                    'serv': {}}
+        command = "rosnode list"
+        try:
+            output = self._app_instance.shell.check_output(command, text=True)
+        except dockerblade.exceptions.CalledProcessError as error:
+            logger.debug("Unable to retrieve rosnode list from command line")
+            raise error
+        node_names = output.split('\r\n')
+        for node_name in node_names:
+            info = f"ros2 node info '{node_name}'"
+            mode: Optional[str] = None
+            try:
+                output = self._app_instance.shell.check_output(info, text=True)
+            except dockerblade.exceptions.CalledProcessError as error:
+                logger.debug(f"Unable to retrieve {info}")
+                raise error
+            output = output.replace(' ', '')
+            lines = output.split('\r\n')
+            for line in lines:
+                if "Publications:" in line:
+                    mode = 'pub'
+                    continue
```

Comment:
This will skip ahead to the next `line` in `lines`. I don't think that's the desired behaviour here. If you want to double check and improve the quality of the codebase, you could write a test case to test this class. (I can give you some pointers on how to do that via Slack, if you want.)

Reply:
There's still at least one bug to be fixed here. Note that `continue` will immediately skip to the next iteration in the loop, and so if any of the if/elif conditions between 53 and 62 are satisfied, that particular line in the file will be ignored.

If the current tests don't reveal the bug, you'll want to refine / add to them to expose this particular bug.

Reply:
Ah, my bad. It looks like this is switching the mode that's used to parse the subsequent lines. It's probably worth adding a comment to the top of this `if/elif` block to explain the logic.